numpy/core/src/umath/ufunc_object.c

```diff
@@ -646,253 +876,141 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
     }
 
     /*
-     * Get keyword output and other arguments.
-     * Raise an error if anything else is present in the
-     * keyword dictionary.
+     * If keywords are present, get keyword output and other arguments.
+     * Raise an error if anything else is present in the keyword dictionary.
      */
-    if (kwds != NULL) {
-        PyObject *key, *value;
-        Py_ssize_t pos = 0;
-        while (PyDict_Next(kwds, &pos, &key, &value)) {
-            Py_ssize_t length = 0;
-            char *str = NULL;
-            int bad_arg = 1;
-
-#if defined(NPY_PY3K)
-            Py_XDECREF(str_key_obj);
-            str_key_obj = PyUnicode_AsASCIIString(key);
-            if (str_key_obj != NULL) {
-                key = str_key_obj;
-            }
-#endif
-
-            if (PyBytes_AsStringAndSize(key, &str, &length) < 0) {
-                PyErr_Clear();
-                PyErr_SetString(PyExc_TypeError, "invalid keyword argument");
-                goto fail;
+    if (kwds == NULL) {
+        return 0; /* no keywords, we're done */
+    }
+    else {
+        PyObject *out_kwd = NULL;
+        PyArray_Descr *dtype = NULL;
+        PyObject *sig = NULL;
+        /*
+         * Parse the keywords, using converters as appropriate.
+         * Note that if any of the out_* is NULL, parse_ufunc_keywords
+         * will treat the keyword as unwanted.
+         */
+        if (parse_ufunc_keywords(
```

Comment:
Can you move the definition of `kwnames` to here?, so that the two things that have to be in the same order are at least adjacent?

Reply:
Yes, makes much more sense.

Reply:
Indeed, I rewrote as
```
static *kwnames[13] = {NULL}
if (kwnames[0] == NULL) {
    kwnames[0] = npy_um_...
...
}
```
so that I don't have the triple `***kwnames` any more. Should be just that little bit faster.